peps/pep-0733.rst

```diff
@@ -0,0 +1,660 @@
+This creates a number of problems.
+
+Bindings for languages other than C/C++ must parse C code
+[`Issue 7 <https://github.com/capi-workgroup/problems/issues/7>`__].
+Some C language features are hard to handle in this way, because
+they produce compiler-dependent output (such as enums) or require
+a C preprocessor/compiler rather than just a parser (such as macros)
+[`Issue 35 <https://github.com/capi-workgroup/problems/issues/35>`__].
+
+Furthermore, C header files tend to expose more than what is intended
+to be part of the public API
+[`Issue 34 <https://github.com/capi-workgroup/problems/issues/34>`__].
+In particular, implementation details such as the precise memory
+layouts of internal data structures can be exposed
+[`Issue 22 <https://github.com/capi-workgroup/problems/issues/22>`__
+and :pep:`620`].
+This can make API evolution very difficult, in particular when it
+occurs in the stable ABI as in the case of ``ob_refcnt`` and ``ob_type``,
+which are accessed via the reference counting macros
+[`Issue 45 <https://github.com/capi-workgroup/problems/issues/45>`__].
+
+We identified a deeper issue in relation to the way that reference
+counting is exposed. The way that C extensions are required to
+manage references with calls to ``Py_INCREF`` and ``Py_DECREF`` is
+specific to CPython's memory model, and is hard for alternative
+Python implementations to emulate.
+[`Issue 12 <https://github.com/capi-workgroup/problems/issues/12>`__].
+
+Another set of problems arises from the fact that a ``PyObject*`` is
+exposed in the C API as an actual pointer rather than a handle. The
+address of an object serves as its ID and is used for comparison,
+and this complicates matters for alternative Python implementations
+that move objects during GC
+[`Issue 37 <https://github.com/capi-workgroup/problems/issues/37>`__].
+
+A separate issue is that object references are opaque to the runtime,
+discoverable only through calls to ``tp_traverse``/``tp_clear``,
+which have their own purposes. If there was a way for the runtime to
+know the structure of the object graph, and keep up with changes in it,
+this would make it possible for alternative implementations to implement
+different memory management schemes
+[`Issue 33 <https://github.com/capi-workgroup/problems/issues/33>`__].
+
+Object Reference Management
+---------------------------
+
+There are C API functions that return borrowed references, and
+functions that steal references to arguments, but there isn't a
+naming convention that makes this obvious, so this is error prone
+[`Issue 8 <https://github.com/capi-workgroup/problems/issues/8>`__
+and `Issue 52 <https://github.com/capi-workgroup/problems/issues/52>`__].
+The terminology used to describe these situations in the documentation
+can also be improved
+[`Issue 11 <https://github.com/capi-workgroup/problems/issues/11>`__].
+
+A more radical change is necessary in the case of functions that
+return borrowed references (such as ``PyList_GetItem``)
```

Comment:
Can we add a section somewhere describing our reference terminology. We are using the same name for different things are it should be made clear.

The term "borrowed" has two different meanings for return values and for arguments.
For arguments, the caller loans a reference to the callee. Perfectly good and safe.
When "returning a borrowed reference", we don't really have a reference to an object at all, just a pointer to memory which we hope is a valid object.
I'd suggest "volatile reference", but this probably isn't the place to be coining new terms.

Reply:
As with steal, maybe `"borrowed"` is an easy way here to use the term we usually use, even though it doesn't really apply?

This is definitely not the place to coin new terms. Plus "volatile" carries as much baggage as "borrowed" (I'm sure it's only a matter of time before someone decides that "volatile" is only allowed to describe unstable people (can I still use "unstable" here?) and we can't use it for technical reasons anymore...), so would also need just as clear a definition.

I'm sure one of the issues somewhere is about clearly defining the terms. That's an important future discussion, but this document should reflect that we currently refer to them as "borrowed".

Reply:
I can just write explicitly:

```
A more radical change is necessary in the case of a function that
returns a ``PyObject*`` that it did not incref (such as ``PyList_GetItem``)
```

Reply:
Incrementing a reference count is an implementation detail. The problem is lack of ownership of the reference.
Can we put "borrow" and "steal" in quotes as @zooba suggests.
We do need a definition somewhere though.

Here's some definitions:
A "stolen" reference is one where the ownership is transferred from caller to callee at the call, permanently.
A "borrowed" parameter reference is where the reference is loaned to the callee by the caller for exactly the duration of the call.
A "borrowed" return reference is not an owned reference at all, just a pointer. We just hope that there exists a reference to the object that will outlive the pointer.

Reply:
I reworded this section. How's this version?